### PR TITLE
launchd: only support setting `command` to a list

### PR DESCRIPTION
**Pull request by Enzime**
@emilazy should I just deprecate setting `command` to a string with a warning?

### CODE REVIEW

**Comment by emilazy** [comment]
Deprecating that seems reasonable, but I’m a bit worried that it means the obvious translation of `prog $a` to `[ "prog" "$a" ]` behaves differently because of the escaping. I guess those cases can use `script` instead? Maybe we should use another name for this variant and just deprecate `command` entirely?

**Comment by Enzime** [comment]
I was thinking of changing the types to any options that set extra arguments to lists as well so that we can run them through `lib.escapeShellArgs` as well

File: modules/launchd/default.nix [hunk 1]
```diff
@@ -80,11 +72,11 @@ let
       };
 
       config = {
-        command = mkIf (config.script != "") (pkgs.writeScript "${name}-start" ''
+        command = mkIf (config.script != "") [ (pkgs.writeScript "${name}-start" ''
```
Comment (emilazy) [inline modules/launchd/default.nix:75]: Should we add a `wait4path` in here? (I’m fine if you’d rather leave that to future work.)

File: modules/services/autossh.nix [hunk 2]
```diff
@@ -91,7 +92,7 @@ in
               # How often AutoSSH checks the network, in seconds
               environment.AUTOSSH_POLL="30";
 
-              command = "${pkgs.autossh}/bin/autossh -M ${toString mport} ${s.extraArguments}";
+              command = [ (lib.getExe pkgs.autossh) "-M" "${toString mport}" ] ++ s.extraArguments;
```
Comment (emilazy) [inline modules/services/autossh.nix:95]: Looks like this is the only place where we pass arbitrary user arguments in this field. After this change users will have to adjust their configurations and wouldn’t be able to reference environment variables here any more. I’m okay with it if you are, but I’m guessing we might have services for which this is undesirable in future.

**Review by emilazy — COMMENTED** [review]
Looks reasonable to me. There’s a test failure; do you know if it’s related?

File: modules/launchd/default.nix (hunk 1 above)
Comment (Enzime) [reply modules/launchd/default.nix:75]: I don't think it's necessary as `serviceConfig.ProgramArguments` already does `wait4path /nix/store`

File: modules/services/autossh.nix (hunk 2 above)
Comment (Enzime) [reply modules/services/autossh.nix:95]: In those situations, I think we should direct them to set `serviceConfig.ProgramArguments` directly (possibly with `lib.mkAfter`)

I did a quick GitHub Code search and didn't see any configs that used environment variables for `services.autossh.extraArguments`:

https://github.com/search?utf8=✓&q=path%3A*.nix+services+autossh+extraArguments&type=Code

**Comment by Enzime** [comment]
Closing this as I don't have the time to keep working on this

**Pull request closed without merging**